File: currency_tracker.go
# currency_tracker.go
/**
 * 💱 Currency Tracker – Live Exchange Rates with ASCII Charts (Go Edition)
 * Advanced: live rates, historical charts, caching, colored output, stats
 */

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strconv"
	"strings"
	"time"
)

// ─── Types ──────────────────────────────────────────────────────────────────

type RateEntry struct {
	Date string  `json:"date"`
	Rate float64 `json:"rate"`
}

type CacheData struct {
	Base       string       `json:"base"`
	Target     string       `json:"target"`
	Rates      []RateEntry  `json:"rates"`
	LastUpdate string       `json:"last_update"`
}

// ─── Configuration ──────────────────────────────────────────────────────────

const (
	defaultBase   = "USD"
	defaultTarget = "EUR"
	historyDays   = 30
	apiBase       = "https://api.exchangerate.host"
)

// ─── Colors ──────────────────────────────────────────────────────────────────

const (
	reset  = "\x1b[0m"
	bright = "\x1b[1m"
	dim    = "\x1b[2m"
	red    = "\x1b[31m"
	green  = "\x1b[32m"
	yellow = "\x1b[33m"
	blue   = "\x1b[34m"
	magenta = "\x1b[35m"
	cyan   = "\x1b[36m"
)

func c(str, color string) string {
	return color + str + reset
}

// ─── API Client ─────────────────────────────────────────────────────────────

type ExchangeRateAPI struct{}

func (api *ExchangeRateAPI) GetLatest(base, target string) (float64, error) {
	url := fmt.Sprintf("%s/latest?base=%s&symbols=%s", apiBase, base, target)
	resp, err := http.Get(url)
	if err != nil {
		return 0, err
	}
	defer resp.Body.Close()
	body, _ := ioutil.ReadAll(resp.Body)
	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return 0, err
	}
	if success, ok := data["success"].(bool); ok && success {
		rates := data["rates"].(map[string]interface{})
		if rate, ok := rates[target].(float64); ok {
			return rate, nil
		}
	}
	return 0, fmt.Errorf("failed to get rate")
}

func (api *ExchangeRateAPI) GetHistory(base, target string, days int) ([]RateEntry, error) {
	end := time.Now()
	start := end.AddDate(0, 0, -days)
	format := func(t time.Time) string { return t.Format("2006-01-02") }
	url := fmt.Sprintf("%s/timeseries?start_date=%s&end_date=%s&base=%s&symbols=%s",
		apiBase, format(start), format(end), base, target)
	resp, err := http.Get(url)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	body, _ := ioutil.ReadAll(resp.Body)
	var data map[string]interface{}
	if err := json.Unmarshal(body, &data); err != nil {
		return nil, err
	}
	if success, ok := data["success"].(bool); ok && success {
		ratesMap := data["rates"].(map[string]interface{})
		var entries []RateEntry
		for date, rateObj := range ratesMap {
			r := rateObj.(map[string]interface{})
			if rate, ok := r[target].(float64); ok {
				entries = append(entries, RateEntry{Date: date, Rate: rate})
			}
		}
		sort.Slice(entries, func(i, j int) bool { return entries[i].Date < entries[j].Date })
		return entries, nil
	}
	return nil, fmt.Errorf("failed to get history")
}

// ─── Cache Manager ─────────────────────────────────────────────────────────

type Cache struct {
	filePath string
	data     CacheData
}

func NewCache() *Cache {
	home, _ := os.UserHomeDir()
	cacheDir := filepath.Join(home, ".currency_tracker")
	os.MkdirAll(cacheDir, 0755)
	c := &Cache{filePath: filepath.Join(cacheDir, "cache.json")}
	c.load()
	return c
}

func (c *Cache) load() {
	if _, err := os.Stat(c.filePath); os.IsNotExist(err) {
		c.data = CacheData{Base: defaultBase, Target: defaultTarget, Rates: []RateEntry{}, LastUpdate: ""}
		return
	}
	raw, err := ioutil.ReadFile(c.filePath)
	if err != nil {
		c.data = CacheData{Base: defaultBase, Target: defaultTarget, Rates: []RateEntry{}, LastUpdate: ""}
		return
	}
	var data CacheData
	if err := json.Unmarshal(raw, &data); err != nil {
		c.data = CacheData{Base: defaultBase, Target: defaultTarget, Rates: []RateEntry{}, LastUpdate: ""}
		return
	}
	c.data = data
}

func (c *Cache) save() {
	raw, _ := json.MarshalIndent(c.data, "", "  ")
	ioutil.WriteFile(c.filePath, raw, 0644)
}

func (c *Cache) Get(base, target string) []RateEntry {
	if c.data.Base == base && c.data.Target == target {
		return c.data.Rates
	}
	return nil
}

func (c *Cache) Set(base, target string, rates []RateEntry) {
	c.data.Base = base
	c.data.Target = target
	c.data.Rates = rates
	c.data.LastUpdate = time.Now().Format(time.RFC3339)
	c.save()
}

func (c *Cache) Clear() {
	c.data = CacheData{Base: defaultBase, Target: defaultTarget, Rates: []RateEntry{}, LastUpdate: ""}
	c.save()
}

// ─── Chart Renderer ─────────────────────────────────────────────────────────

func drawAsciiChart(rates []RateEntry, width, height int) string {
	if len(rates) == 0 {
		return "No data available."
	}
	values := make([]float64, len(rates))
	for i, r := range rates {
		values[i] = r.Rate
	}
	minVal, maxVal := values[0], values[0]
	for _, v := range values {
		if v < minVal {
			minVal = v
		}
		if v > maxVal {
			maxVal = v
		}
	}
	rangeVal := maxVal - minVal
	if rangeVal == 0 {
		return fmt.Sprintf("Rate is constant at %.4f", values[0])
	}
	normalized := make([]int, len(values))
	for i, v := range values {
		normalized[i] = int(math.Floor((v - minVal) / rangeVal * float64(height-1)))
	}
	var lines []string
	for row := height - 1; row >= 0; row-- {
		var line strings.Builder
		for i, val := range normalized {
			if val >= row {
				if i > 0 && normalized[i-1] >= row {
					line.WriteString("─")
				} else {
					line.WriteString("┌")
				}
			} else {
				line.WriteString(" ")
			}
		}
		if strings.TrimSpace(line.String()) != "" {
			lines = append(lines, line.String())
		}
	}
	// X axis
	step := len(rates) / 10
	if step < 1 {
		step = 1
	}
	var xAxis strings.Builder
	xAxis.WriteString(" ")
	lastPos := 0
	for i := 0; i < len(rates); i += step {
		label := rates[i].Date[5:10]
		spaces := i - lastPos
		if spaces > 0 {
			xAxis.WriteString(strings.Repeat(" ", spaces))
		}
		xAxis.WriteString(label)
		lastPos = i
	}
	if lastPos < len(rates)-1 {
		xAxis.WriteString(strings.Repeat(" ", len(rates)-1-lastPos))
	}
	result := strings.Join(lines, "\n") + "\n" + xAxis.String()
	result += fmt.Sprintf("\nMin: %.4f  Max: %.4f", minVal, maxVal)
	return result
}

// ─── Main App ──────────────────────────────────────────────────────────────

type CurrencyTracker struct {
	api    *ExchangeRateAPI
	cache  *Cache
	base   string
	target string
	rates  []RateEntry
	reader *bufio.Reader
}

func NewCurrencyTracker() *CurrencyTracker {
	t := &CurrencyTracker{
		api:    &ExchangeRateAPI{},
		cache:  NewCache(),
		base:   defaultBase,
		target: defaultTarget,
		reader: bufio.NewReader(os.Stdin),
	}
	t.loadCachedOrFetch()
	return t
}

func (t *CurrencyTracker) loadCachedOrFetch() {
	if cached := t.cache.Get(t.base, t.target); cached != nil {
		t.rates = cached
		fmt.Println(c(fmt.Sprintf("📂 Loaded cached data for %s/%s", t.base, t.target), dim))
	} else {
		t.fetchHistory()
	}
}

func (t *CurrencyTracker) fetchHistory() {
	fmt.Print(c("Fetching historical data...", dim))
	rates, err := t.api.GetHistory(t.base, t.target, historyDays)
	if err != nil || len(rates) == 0 {
		fmt.Println(c(" ⚠️  No historical data retrieved.", yellow))
		return
	}
	t.rates = rates
	t.cache.Set(t.base, t.target, rates)
	fmt.Println(c(fmt.Sprintf(" ✅ Retrieved %d data points", len(rates)), green))
}

func (t *CurrencyTracker) getCurrentRate() (float64, error) {
	rate, err := t.api.GetLatest(t.base, t.target)
	if err == nil {
		return rate, nil
	}
	if len(t.rates) > 0 {
		return t.rates[len(t.rates)-1].Rate, nil
	}
	return 0, fmt.Errorf("no rate available")
}

func (t *CurrencyTracker) ask(prompt string) string {
	fmt.Print(prompt)
	line, _ := t.reader.ReadString('\n')
	return strings.TrimSpace(line)
}

func (t *CurrencyTracker) askConfirm(prompt string) bool {
	ans := t.ask(prompt + " (yes/no): ")
	ans = strings.ToLower(ans)
	return ans == "yes" || ans == "y"
}

func (t *CurrencyTracker) showMenu() {
	current, _ := t.getCurrentRate()
	currentStr := "N/A"
	if current != 0 {
		currentStr = fmt.Sprintf("%.4f", current)
	}
	fmt.Println("\n" + c(strings.Repeat("═", 50), cyan))
	fmt.Println(c("💱 CURRENCY TRACKER", bright+cyan))
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Printf("  Base: %s  Target: %s\n", t.base, t.target)
	fmt.Printf("  Current: %s\n", currentStr)
	fmt.Printf("  Data points: %d\n", len(t.rates))
	fmt.Println(c(strings.Repeat("═", 50), cyan))
	fmt.Println("  1. 📈 Show chart")
	fmt.Println("  2. 💹 Show current rate")
	fmt.Println("  3. 🔄 Change currency pair")
	fmt.Println("  4. 📊 Show statistics")
	fmt.Println("  5. 🔄 Refresh data")
	fmt.Println("  6. 🗑️  Clear cache")
	fmt.Println("  0. 🚪 Exit")
	fmt.Println(c(strings.Repeat("═", 50), cyan))
}

func (t *CurrencyTracker) showChart() {
	if len(t.rates) == 0 {
		fmt.Println(c("No data available. Fetching...", yellow))
		t.fetchHistory()
	}
	if len(t.rates) == 0 {
		fmt.Println(c("No data to display.", red))
		return
	}
	fmt.Println(c(fmt.Sprintf("\n📈 Exchange Rate Chart (last %d days)", len(t.rates)), bright))
	fmt.Println(drawAsciiChart(t.rates, 50, 10))
	fmt.Println(c(fmt.Sprintf("\nLatest: %.4f", t.rates[len(t.rates)-1].Rate), cyan))
}

func (t *CurrencyTracker) showRate() {
	rate, err := t.getCurrentRate()
	if err == nil {
		fmt.Println(c(fmt.Sprintf("\n💹 Current %s/%s: %.4f", t.base, t.target, rate), green))
	} else {
		fmt.Println(c("Could not fetch rate.", red))
	}
}

func (t *CurrencyTracker) showStats() {
	if len(t.rates) == 0 {
		fmt.Println(c("No data available.", yellow))
		return
	}
	var sum float64
	minVal := t.rates[0].Rate
	maxVal := t.rates[0].Rate
	for _, r := range t.rates {
		sum += r.Rate
		if r.Rate < minVal {
			minVal = r.Rate
		}
		if r.Rate > maxVal {
			maxVal = r.Rate
		}
	}
	avg := sum / float64(len(t.rates))
	volatility := (maxVal - minVal) / avg * 100
	last := t.rates[len(t.rates)-1].Rate
	fmt.Println("\n📊 STATISTICS")
	fmt.Println(c(strings.Repeat("─", 30), dim))
	fmt.Printf("  Period:      %d days\n", len(t.rates))
	fmt.Printf("  Average:     %.4f\n", avg)
	fmt.Printf("  Min:         %.4f\n", minVal)
	fmt.Printf("  Max:         %.4f\n", maxVal)
	fmt.Printf("  Volatility:  %.2f%%\n", volatility)
	fmt.Printf("  Current:     %.4f\n", last)
}

func (t *CurrencyTracker) changePair() {
	base := t.ask(fmt.Sprintf("Base currency (default %s): ", t.base))
	target := t.ask(fmt.Sprintf("Target currency (default %s): ", t.target))
	newBase := strings.ToUpper(base)
	if newBase == "" {
		newBase = t.base
	}
	newTarget := strings.ToUpper(target)
	if newTarget == "" {
		newTarget = t.target
	}
	if newBase == t.base && newTarget == t.target {
		fmt.Println("Pair unchanged.")
		return
	}
	t.base = newBase
	t.target = newTarget
	if cached := t.cache.Get(t.base, t.target); cached != nil {
		t.rates = cached
		fmt.Println(c(fmt.Sprintf("📂 Loaded cached data for %s/%s", t.base, t.target), dim))
	} else {
		t.fetchHistory()
	}
}

func (t *CurrencyTracker) refresh() {
	t.fetchHistory()
	fmt.Println(c("✅ Data refreshed.", green))
}

func (t *CurrencyTracker) clearCache() {
	if !t.askConfirm("🗑️  Delete all cached data?") {
		return
	}
	t.cache.Clear()
	t.rates = []RateEntry{}
	fmt.Println(c("Cache cleared.", yellow))
}

func (t *CurrencyTracker) run() {
	fmt.Print("\033[H\033[2J")
	fmt.Println(c("\n💱 Currency Tracker – Live Exchange Rates with Charts", bright+cyan))
	fmt.Println(c("Track the pulse of the global economy.", dim))

	for {
		t.showMenu()
		choice := t.ask("Your choice: ")
		switch choice {
		case "1":
			t.showChart()
		case "2":
			t.showRate()
		case "3":
			t.changePair()
		case "4":
			t.showStats()
		case "5":
			t.refresh()
		case "6":
			t.clearCache()
		case "0":
			fmt.Println(c("👋 Goodbye!", cyan))
			return
		default:
			fmt.Println(c("❌ Invalid choice.", red))
		}
		if choice != "0" {
			fmt.Print("\nPress Enter to continue...")
			t.reader.ReadString('\n')
		}
	}
}

// ─── Main ────────────────────────────────────────────────────────────────────

func main() {
	app := NewCurrencyTracker()
	app.run()
}

File: currency_tracker.js
# currency_tracker.js
/**
 * 💱 Currency Tracker – Live Exchange Rates with ASCII Charts (Node.js Edition)
 * Advanced: live rates, historical charts, caching, colored output, stats
 */

const fs = require('fs');
const path = require('path');
const os = require('os');
const readline = require('readline');
const fetch = require('node-fetch');

// ─── Configuration ──────────────────────────────────────────────────────────

const CONFIG = {
    cacheDir: path.join(os.homedir(), '.currency_tracker'),
    cacheFile: 'cache.json',
    apiBase: 'https://api.exchangerate.host',
    defaultBase: 'USD',
    defaultTarget: 'EUR',
    historyDays: 30,
};

// ─── Colors ──────────────────────────────────────────────────────────────────

const colors = {
    reset: '\x1b[0m',
    bright: '\x1b[1m',
    dim: '\x1b[2m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    magenta: '\x1b[35m',
    cyan: '\x1b[36m',
};

const c = (str, color) => `${color}${str}${colors.reset}`;

// ─── API Client ─────────────────────────────────────────────────────────────

class ExchangeRateAPI {
    static async getLatest(base, target) {
        const url = `${CONFIG.apiBase}/latest?base=${base}&symbols=${target}`;
        try {
            const resp = await fetch(url);
            const data = await resp.json();
            if (data.success) return data.rates[target];
            return null;
        } catch (e) {
            console.error(c(`❌ API error: ${e.message}`, colors.red));
            return null;
        }
    }

    static async getHistory(base, target, days = CONFIG.historyDays) {
        const end = new Date();
        const start = new Date();
        start.setDate(end.getDate() - days);
        const format = d => d.toISOString().slice(0, 10);
        const url = `${CONFIG.apiBase}/timeseries?start_date=${format(start)}&end_date=${format(end)}&base=${base}&symbols=${target}`;
        try {
            const resp = await fetch(url);
            const data = await resp.json();
            if (data.success) {
                const rates = data.rates;
                const entries = Object.entries(rates)
                    .sort((a, b) => a[0].localeCompare(b[0]))
                    .map(([date, ratesObj]) => [date, ratesObj[target]]);
                return entries;
            }
            return [];
        } catch (e) {
            console.error(c(`❌ API error: ${e.message}`, colors.red));
            return [];
        }
    }
}

// ─── Cache Manager ──────────────────────────────────────────────────────────

class Cache {
    constructor() {
        if (!fs.existsSync(CONFIG.cacheDir)) fs.mkdirSync(CONFIG.cacheDir, { recursive: true });
        this.filePath = path.join(CONFIG.cacheDir, CONFIG.cacheFile);
        this.data = this._load();
    }

    _load() {
        if (fs.existsSync(this.filePath)) {
            try {
                return JSON.parse(fs.readFileSync(this.filePath, 'utf8'));
            } catch (_) { /* ignore */ }
        }
        return { base: CONFIG.defaultBase, target: CONFIG.defaultTarget, rates: [], last_update: '' };
    }

    save() {
        fs.writeFileSync(this.filePath, JSON.stringify(this.data, null, 2));
    }

    get(base, target) {
        if (this.data.base === base && this.data.target === target) {
            return this.data.rates.map(r => [r.date, r.rate]);
        }
        return null;
    }

    set(base, target, rates) {
        this.data.base = base;
        this.data.target = target;
        this.data.rates = rates.map(([date, rate]) => ({ date, rate }));
        this.data.last_update = new Date().toISOString();
        this.save();
    }

    clear() {
        this.data = { base: CONFIG.defaultBase, target: CONFIG.defaultTarget, rates: [], last_update: '' };
        this.save();
    }
}

// ─── Chart Renderer ─────────────────────────────────────────────────────────

function drawAsciiChart(rates, width = 50, height = 10) {
    if (!rates || rates.length === 0) return 'No data available.';

    const values = rates.map(([_, r]) => r);
    const minVal = Math.min(...values);
    const maxVal = Math.max(...values);
    const range = maxVal - minVal;
    if (range === 0) return `Rate is constant at ${values[0].toFixed(4)}`;

    const normalized = values.map(v => Math.floor((v - minVal) / range * (height - 1)));
    const lines = [];
    for (let row = height - 1; row >= 0; row--) {
        let line = '';
        for (let i = 0; i < normalized.length; i++) {
            if (normalized[i] >= row) {
                if (i > 0 && normalized[i-1] >= row) line += '─';
                else line += '┌';
            } else {
                line += ' ';
            }
        }
        if (line.trim() !== '') lines.push(line);
    }
    // X axis labels
    const step = Math.max(1, Math.floor(rates.length / 10));
    let xAxis = ' ';
    let lastPos = 0;
    for (let i = 0; i < rates.length; i += step) {
        const label = rates[i][0].slice(5, 10); // MM-DD
        xAxis += ' '.repeat(i - lastPos) + label;
        lastPos = i;
    }
    if (lastPos < rates.length - 1) {
        xAxis += ' '.repeat(rates.length - 1 - lastPos);
    }
    return lines.join('\n') + '\n' + xAxis + `\nMin: ${minVal.toFixed(4)}  Max: ${maxVal.toFixed(4)}`;
}

// ─── Main App ──────────────────────────────────────────────────────────────

class CurrencyTracker {
    constructor() {
        this.rl = readline.createInterface({ input: process.stdin, output: process.stdout });
        this.cache = new Cache();
        this.base = CONFIG.defaultBase;
        this.target = CONFIG.defaultTarget;
        this.rates = [];
        this._loadCachedOrFetch();
    }

    async _loadCachedOrFetch() {
        const cached = this.cache.get(this.base, this.target);
        if (cached) {
            this.rates = cached;
            console.log(c(`📂 Loaded cached data for ${this.base}/${this.target}`, colors.dim));
        } else {
            await this._fetchHistory();
        }
    }

    async _fetchHistory() {
        process.stdout.write(c('Fetching historical data...', colors.dim));
        this.rates = await ExchangeRateAPI.getHistory(this.base, this.target);
        if (this.rates && this.rates.length) {
            this.cache.set(this.base, this.target, this.rates);
            console.log(c(` ✅ Retrieved ${this.rates.length} data points`, colors.green));
        } else {
            console.log(c(' ⚠️  No historical data retrieved. Using cache if available.', colors.yellow));
        }
    }

    async _getCurrentRate() {
        const rate = await ExchangeRateAPI.getLatest(this.base, this.target);
        if (rate !== null) return rate;
        if (this.rates.length) return this.rates[this.rates.length-1][1];
        return null;
    }

    _ask(prompt) {
        return new Promise(resolve => this.rl.question(prompt, resolve));
    }

    async _askConfirm(prompt) {
        const ans = await this._ask(prompt + ' (yes/no): ');
        return ans.trim().toLowerCase() === 'yes' || ans.trim().toLowerCase() === 'y';
    }

    async showMenu() {
        const current = await this._getCurrentRate() || 'N/A';
        console.log('\n' + c('═'.repeat(50), colors.cyan));
        console.log(c('💱 CURRENCY TRACKER', colors.bright + colors.cyan));
        console.log(c('═'.repeat(50), colors.cyan));
        console.log(`  Base: ${this.base}  Target: ${this.target}`);
        console.log(`  Current: ${typeof current === 'number' ? current.toFixed(4) : current}`);
        console.log(`  Data points: ${this.rates.length}`);
        console.log(c('═'.repeat(50), colors.cyan));
        console.log('  1. 📈 Show chart');
        console.log('  2. 💹 Show current rate');
        console.log('  3. 🔄 Change currency pair');
        console.log('  4. 📊 Show statistics');
        console.log('  5. 🔄 Refresh data');
        console.log('  6. 🗑️  Clear cache');
        console.log('  0. 🚪 Exit');
        console.log(c('═'.repeat(50), colors.cyan));
    }

    showChart() {
        if (!this.rates || this.rates.length === 0) {
            console.log(c('No data available. Fetching...', colors.yellow));
            return;
        }
        console.log(c(`\n📈 Exchange Rate Chart (last ${this.rates.length} days)`, colors.bright));
        console.log(drawAsciiChart(this.rates));
        console.log(c(`\nLatest: ${this.rates[this.rates.length-1][1].toFixed(4)}`, colors.cyan));
    }

    async showRate() {
        const rate = await this._getCurrentRate();
        if (rate !== null) {
            console.log(c(`\n💹 Current ${this.base}/${this.target}: ${rate.toFixed(4)}`, colors.green));
        } else {
            console.log(c('Could not fetch rate.', colors.red));
        }
    }

    showStats() {
        if (!this.rates || this.rates.length === 0) {
            console.log(c('No data available.', colors.yellow));
            return;
        }
        const values = this.rates.map(([_, r]) => r);
        const avg = values.reduce((a, b) => a + b, 0) / values.length;
        const minv = Math.min(...values);
        const maxv = Math.max(...values);
        const volatility = (maxv - minv) / avg * 100;
        const last = values[values.length-1];
        console.log('\n📊 STATISTICS');
        console.log(c('─'.repeat(30), colors.dim));
        console.log(`  Period:      ${this.rates.length} days`);
        console.log(`  Average:     ${avg.toFixed(4)}`);
        console.log(`  Min:         ${minv.toFixed(4)}`);
        console.log(`  Max:         ${maxv.toFixed(4)}`);
        console.log(`  Volatility:  ${volatility.toFixed(2)}%`);
        console.log(`  Current:     ${last.toFixed(4)}`);
    }

    async changePair() {
        const base = await this._ask(`Base currency (default ${this.base}): `);
        const target = await this._ask(`Target currency (default ${this.target}): `);
        const newBase = base.trim().toUpperCase() || this.base;
        const newTarget = target.trim().toUpperCase() || this.target;
        if (newBase === this.base && newTarget === this.target) {
            console.log('Pair unchanged.');
            return;
        }
        this.base = newBase;
        this.target = newTarget;
        const cached = this.cache.get(this.base, this.target);
        if (cached) {
            this.rates = cached;
            console.log(c(`📂 Loaded cached data for ${this.base}/${this.target}`, colors.dim));
        } else {
            await this._fetchHistory();
        }
    }

    async refresh() {
        await this._fetchHistory();
        console.log(c('✅ Data refreshed.', colors.green));
    }

    async clearCache() {
        const ok = await this._askConfirm('🗑️  Delete all cached data?');
        if (!ok) return;
        this.cache.clear();
        this.rates = [];
        console.log(c('Cache cleared.', colors.yellow));
    }

    async run() {
        console.clear();
        console.log(c('\n💱 Currency Tracker – Live Exchange Rates with Charts', colors.bright + colors.cyan));
        console.log(c('Track the pulse of the global economy.', colors.dim));

        while (true) {
            await this.showMenu();
            const choice = await this._ask('Your choice: ');
            switch (choice.trim()) {
                case '1': this.showChart(); break;
                case '2': await this.showRate(); break;
                case '3': await this.changePair(); break;
                case '4': this.showStats(); break;
                case '5': await this.refresh(); break;
                case '6': await this.clearCache(); break;
                case '0':
                    console.log(c('👋 Goodbye!', colors.cyan));
                    this.rl.close();
                    return;
                default:
                    console.log(c('❌ Invalid choice.', colors.red));
            }
            if (choice !== '0') {
                console.log('\nPress Enter to continue...');
                await this._ask('');
            }
        }
    }
}

// ─── Main ────────────────────────────────────────────────────────────────────

const main = async () => {
    try {
        const app = new CurrencyTracker();
        await app.run();
    } catch (e) {
        console.error(c(`❌ Unexpected error: ${e.message}`, colors.red));
        process.exit(1);
    }
};

main();
